### PR TITLE
GH-93897: Store frame size in code object and de-opt if insufficient space on thread frame stack.

### PR DESCRIPTION
**Pull request by markshannon**
Speeds up pushing a frame, by removing a few branches.

* De-opt if there isn't space for a frame. This should be very rare.
* Move check for NULL locals off fast path.
* Compute size of frame when creating code object, not when pushing frame.

### CODE REVIEW

File: Include/internal/pycore_frame.h [hunk 1]
```diff
@@ -94,20 +95,20 @@ static inline void _PyFrame_StackPush(_PyInterpreterFrame *f, PyObject *value) {
...
     _PyInterpreterFrame *frame, PyFunctionObject *func,
-    PyObject *locals, int nlocalsplus)
+    PyObject *locals, PyCodeObject *code)
 {
     frame->f_func = func;
-    frame->f_code = (PyCodeObject *)Py_NewRef(func->func_code);
+    frame->f_code = (PyCodeObject *)Py_NewRef(code);
     frame->f_builtins = func->func_builtins;
     frame->f_globals = func->func_globals;
-    frame->f_locals = Py_XNewRef(locals);
```
Comment (Fidget-Spinner) [inline Include/internal/pycore_frame.h:107]: Out of interest, why did you move to incref to out of this function? Does this mean there are some situations where the caller already holds a strong reference that's meant to be transferred?

**Review by Fidget-Spinner — APPROVED** [review]
LGTM in general, I just have one question.

File: Include/internal/pycore_frame.h (hunk 1 above)
Comment (markshannon) [reply Include/internal/pycore_frame.h:107]: This now consumes a reference to `locals`, so any caller passing in non-NULL needs to allow for that.
Since `locals` is almost always `NULL`, this seems more efficient.

Comment (Fidget-Spinner) [reply Include/internal/pycore_frame.h:107]: Ahh ok, so it removes a branch in most cases too (from the `Py_XNewRef`). Cool.
